gearcmd/worker.go

```diff
@@ -303,34 +304,54 @@ func (conf *TaskConfig) doProcess(job baseworker.Job, envVars []string, tryCount
 	}
 }
 
-// stopProcess kills a given process. It's second argument is a grace period, if > 0, SIGTERM will
-// be sent first folloed by SIGKILL after the grace period is over. If the grace period is 0, this
-// is simply a hard SIGKILL.
+// stopProcess kills a given process. It's second argument is a grace period.
+// If, after the grace period, the process hasn't exited, SIGKILL will be sent.
+// It also calls os.Exit, since we currently rely on cutting off the connection
+// with gearmand to trigger reassignment of work to another worker.
 func stopProcess(p *os.Process, gracePeriod time.Duration) error {
 	lg.InfoD("stopping-process", logger.M{"pid": p.Pid, "grace_period": gracePeriod})
-	if gracePeriod > 0 {
-		// we use SIGTERM so that the subprocess can gracefully exit
-		if err := syscall.Kill(p.Pid, syscall.SIGTERM); err != nil {
-			return fmt.Errorf("unable to send SIGTERM, error: %s", err)
-		}
-		time.Sleep(gracePeriod)
+	if err := p.Signal(os.Signal(syscall.SIGTERM)); err != nil {
+		return fmt.Errorf("unable to send SIGTERM, error: %s", err)
 	}
-	// kill entire group of process spawned by our cmd.Process
-	targetID := p.Pid
-	pgid, err := syscall.Getpgid(p.Pid)
+	timer := time.AfterFunc(gracePeriod, func() {
+		// kill entire group of process spawned by our cmd.Process
+		targetID := p.Pid
+		pgid, err := syscall.Getpgid(p.Pid)
+		if err != nil {
+			lg.InfoD("unable-to-get-pgid", logger.M{"pid": p.Pid})
+		} else {
+			// minus sign required to kill PGIDs
+			// https://linux.die.net/man/2/kill
+			targetID = -pgid
+		}
+		lg.InfoD("killing-pid", logger.M{"pid": p.Pid, "target_id": targetID})
+		syscall.Kill(targetID, syscall.SIGKILL)
```

Comment:
explicitly state here why we are ignoring the error returned by `syscall.Kill`

Reply:
a60ca91